perception/lidar_transfusion/lib/network/network_trt.cpp

```diff
@@ -0,0 +1,231 @@
+    tensorrt_common::LOG_ERROR(logger_) << "Failed to create config" << std::endl;
+    return false;
+  }
+#if (NV_TENSORRT_MAJOR * 1000) + (NV_TENSORRT_MINOR * 100) + NV_TENSOR_PATCH >= 8400
+  config->setMemoryPoolLimit(nvinfer1::MemoryPoolType::kWORKSPACE, workspace_size);
+#else
+  config->setMaxWorkspaceSize(workspace_size);
+#endif
+  if (precision == "fp16") {
+    if (builder->platformHasFastFp16()) {
+      tensorrt_common::LOG_INFO(logger_) << "Using TensorRT FP16 Inference" << std::endl;
+      config->setFlag(nvinfer1::BuilderFlag::kFP16);
+    } else {
+      tensorrt_common::LOG_INFO(logger_)
+        << "TensorRT FP16 Inference isn't supported in this environment" << std::endl;
+    }
+  }
+
+  const auto flag =
+    1U << static_cast<uint32_t>(nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH);
+  auto network =
+    tensorrt_common::TrtUniquePtr<nvinfer1::INetworkDefinition>(builder->createNetworkV2(flag));
+  if (!network) {
+    tensorrt_common::LOG_ERROR(logger_) << "Failed to create network" << std::endl;
+    return false;
+  }
+
+  auto parser = tensorrt_common::TrtUniquePtr<nvonnxparser::IParser>(
+    nvonnxparser::createParser(*network, logger_));
+  parser->parseFromFile(onnx_path.c_str(), static_cast<int>(nvinfer1::ILogger::Severity::kERROR));
+
+  if (!setProfile(*builder, *network, *config)) {
+    tensorrt_common::LOG_ERROR(logger_) << "Failed to set profile" << std::endl;
+    return false;
+  }
+
+  plan_ = tensorrt_common::TrtUniquePtr<nvinfer1::IHostMemory>(
+    builder->buildSerializedNetwork(*network, *config));
+  if (!plan_) {
+    tensorrt_common::LOG_ERROR(logger_) << "Failed to create serialized network" << std::endl;
+    return false;
+  }
+  engine = tensorrt_common::TrtUniquePtr<nvinfer1::ICudaEngine>(
+    runtime_->deserializeCudaEngine(plan_->data(), plan_->size()));
+  if (!engine) {
+    tensorrt_common::LOG_ERROR(logger_) << "Failed to create engine" << std::endl;
+    return false;
+  }
+
+  return saveEngine(engine_path);
+}
+
+bool NetworkTRT::saveEngine(const std::string & engine_path)
+{
+  tensorrt_common::LOG_INFO(logger_) << "Writing to " << engine_path << std::endl;
+  std::ofstream file(engine_path, std::ios::out | std::ios::binary);
+  file.write(reinterpret_cast<const char *>(plan_->data()), plan_->size());
+  return validateNetworkIO();
+}
+
+bool NetworkTRT::loadEngine(const std::string & engine_path)
+{
+  std::ifstream engine_file(engine_path);
+  std::stringstream engine_buffer;
+  engine_buffer << engine_file.rdbuf();
+  std::string engine_str = engine_buffer.str();
+  engine = tensorrt_common::TrtUniquePtr<nvinfer1::ICudaEngine>(runtime_->deserializeCudaEngine(
+    reinterpret_cast<const void *>(engine_str.data()), engine_str.size()));
+  tensorrt_common::LOG_INFO(logger_) << "Loaded engine from " << engine_path << std::endl;
+  return validateNetworkIO();
+}
+
+bool NetworkTRT::validateNetworkIO()
+{
+  if (engine->getNbIOTensors() != NetworkIO::ENUM_SIZE) {
```

Comment:
Thank your creating some checks (I think we talked about this a while ago).
Although the number of I/O is a good one, it may also be useful to check the dimensionality of some of the tensors.
For example, in centerpoint there is:

```
  for (std::size_t ci = 0; ci < out_channel_sizes_.size(); ci++) {
    auto out_name = network.getOutput(ci)->getName();

    if (
      out_name == std::string("heatmap") &&
      network.getOutput(ci)->getDimensions().d[1] != static_cast<int32_t>(out_channel_sizes_[ci])) {
      tensorrt_common::LOG_ERROR(logger_)
        << "Expected and actual number of classes do not match" << std::endl;
      return false;
    }
    auto out_dims = nvinfer1::Dims4(
      config_.batch_size_, out_channel_sizes_[ci], config_.down_grid_size_y_,
      config_.down_grid_size_x_);
    profile->setDimensions(out_name, nvinfer1::OptProfileSelector::kMIN, out_dims);
    profile->setDimensions(out_name, nvinfer1::OptProfileSelector::kOPT, out_dims);
    profile->setDimensions(out_name, nvinfer1::OptProfileSelector::kMAX, out_dims);
  }
```

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/e83d0bffa1142767c7b968b4661a140390f2e34a
Agree.
First, the optimization profile has to match that one in TensorRT engine. It prevents a scenario when user changed the config but forgot to delete previous engine file.
Secondly the tensors shapes have to match actual network. The `lidar_transfusion` handles inference for specific network configuration. Some parameters are exposed, but most of them are defined in header file. It prevents a scenario when user provided custom network model which may not follow our assumptions (e.g. num of classes).

Reply:
Thanks ! :star_struck: